In [1]:
%pip install PyPDF2 pdfplumber


     -------------------------------------- 43.7/43.7 kB 429.2 kB/s eta 0:00:00
     -------------------------------------- 143.6/143.6 kB 1.7 MB/s eta 0:00:00
     ---------------------------------------- 5.6/5.6 MB 5.2 MB/s eta 0:00:00


  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [7]:
import PyPDF2

def extract_text_pypdf2(file_path):
    with open(file_path, 'rb') as file:
        pdf_reader = PyPDF2.PdfReader(file)
        text = ""
        print(len(pdf_reader.pages))
        for page_num in range(len(pdf_reader.pages)):
            page = pdf_reader.getPage(page_num)
            text += page.extractText()

    return text

file_path = "sample.pdf"
extracted_text = extract_text_pypdf2(file_path)
print(extracted_text)


1


DeprecationError: reader.getPage(pageNumber) is deprecated and was removed in PyPDF2 3.0.0. Use reader.pages[page_number] instead.

In [8]:
import pdfplumber

def extract_text_pdfplumber(file_path):
    with pdfplumber.open(file_path) as pdf:
        text = ""

        for page in pdf.pages:
            text += page.extract_text()

    return text

file_path = "sample.pdf"
extracted_text = extract_text_pdfplumber(file_path)
print(extracted_text)


Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been
the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of
type and scrambled it to make a type specimen book. It has survived not only five centuries, but also
the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the
1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with
desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been
the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of
type and scrambled it to make a type specimen book. It has survived not only five centuries, but also
the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the
1960s with the release of Letras

In [9]:
%pip install nltk spacy

     ---------------------------------------- 1.5/1.5 MB 3.3 MB/s eta 0:00:00
     ---------------------------------------- 12.6/12.6 MB 4.1 MB/s eta 0:00:00
     -------------------------------------- 298.0/298.0 kB 9.3 MB/s eta 0:00:00
     ---------------------------------------- 96.6/96.6 kB 5.4 MB/s eta 0:00:00
     ---------------------------------------- 96.7/96.7 kB 5.8 MB/s eta 0:00:00
     ------------------------------------- 482.7/482.7 kB 10.1 MB/s eta 0:00:00
     ---------------------------------------- 56.8/56.8 kB 3.1 MB/s eta 0:00:00
     ---------------------------------------- 48.9/48.9 kB 2.4 MB/s eta 0:00:00
     ------------------------------------- 181.6/181.6 kB 10.7 MB/s eta 0:00:00
     ---------------------------------------- 1.5/1.5 MB 7.3 MB/s eta 0:00:00
     -------------------------------------- 133.1/133.1 kB 7.7 MB/s eta 0:00:00
     ---------------------------------------- 7.0/7.0 MB 7.9 MB/s eta 0:00:00
  Attempting uninstall: colorama
    Found exi

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [13]:
!python -m spacy download en_core_web_sm

     ---------------------------------------- 12.8/12.8 MB 7.3 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [11]:
import nltk
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
import string

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

def preprocess_text_nltk(text):
    # Tokenize words and sentences
    words = word_tokenize(text)
    sentences = sent_tokenize(text)

    # Remove punctuation and lowercase words
    words = [word.lower() for word in words if word not in string.punctuation]

    # Remove stopwords
    stop_words = set(stopwords.words('english'))
    words = [word for word in words if word not in stop_words]

    # Lemmatize words
    lemmatizer = WordNetLemmatizer()
    words = [lemmatizer.lemmatize(word) for word in words]

    return words, sentences

text = "Here is some sample text for preprocessing with NLTK. Let's see the result!"
processed_words, processed_sentences = preprocess_text_nltk(text)
print(processed_words)
print(processed_sentences)


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\souvik.dey\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\souvik.dey\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\souvik.dey\AppData\Roaming\nltk_data...


['sample', 'text', 'preprocessing', 'nltk', 'let', "'s", 'see', 'result']
['Here is some sample text for preprocessing with NLTK.', "Let's see the result!"]


In [16]:
import spacy

nlp = spacy.load("en_core_web_sm")
doc = nlp("Apple is looking at buying U.K. startup for $1 billion")
for token in doc:
    print(token.text, token.pos_, token.dep_)

Apple PROPN nsubj
is AUX aux
looking VERB ROOT
at ADP prep
buying VERB pcomp
U.K. PROPN dobj
startup NOUN dep
for ADP prep
$ SYM quantmod
1 NUM compound
billion NUM pobj


In [18]:
# import spacy

# nlp = spacy.load("en_core_web_sm")

# def preprocess_text_spacy(text):
#     # Process text with spaCy
#     doc = nlp(text)

#     # Tokenize words and sentences
#     words = [token.text for token in doc]
#     sentences = [sent.text for sent in doc.sents]

#     # Remove punctuation and lowercase words
#     words = [word.lower() for word in words if not word.is_punct]

#     # Remove stopwords
#     words = [word for word in words if not nlp.vocab[word].is_stop]

#     # Lemmatize words
#     words = [token.lemma_ for token in doc if not token.is_punct and not token.is_stop]

#     return words, sentences

# text = "Here is some sample text for preprocessing with spaCy. Let's see the result!"
# processed_words, processed_sentences = preprocess_text_spacy(text)
# print(processed_words)
# print(processed_sentences)

import spacy

nlp = spacy.load("en_core_web_sm")

def preprocess_text_spacy(text):
    # Process text with spaCy
    doc = nlp(text)

    # Tokenize words and sentences
    words = [token.text for token in doc if not token.is_punct]
    sentences = [sent.text for sent in doc.sents]

    # Remove punctuation and lowercase words
    words = [word.lower() for word in words]

    # Remove stopwords
    words = [word for word, token in zip(words, doc) if not token.is_stop]

    # Lemmatize words
    words = [token.lemma_ for token in doc if not token.is_punct and not token.is_stop]

    return words, sentences

text = "Here is some sample text for preprocessing with spaCy. Let's see the result!"
processed_words, processed_sentences = preprocess_text_spacy(text)
print(processed_words)
print(processed_sentences)



['sample', 'text', 'preprocesse', 'spaCy', 'let', 'result']
['Here is some sample text for preprocessing with spaCy.', "Let's see the result!"]


In [20]:
%pip install gtts

Note: you may need to restart the kernel to use updated packages.


  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [23]:
from gtts import gTTS

def text_to_speech_gtts(text, output_file):
    tts = gTTS(text, lang='en')
    tts.save(output_file)

#text = "This is a sample text for converting to speech using gTTS."
text = "Here is some sample text for preprocessing with spaCy! Let's see the result!"
output_file = "output_gtts.mp3"
text_to_speech_gtts(text, output_file)


In [24]:
%pip install pyttsx3



     -------------------------------------- 172.8/172.8 kB 1.3 MB/s eta 0:00:00


In [25]:
import pyttsx3

def text_to_speech_pyttsx3(text, output_file):
    engine = pyttsx3.init()
    engine.save_to_file(text, output_file)
    engine.runAndWait()

text = "This is a sample text for converting to speech using pyttsx3."
output_file = "output_pyttsx3.mp3"
text_to_speech_pyttsx3(text, output_file)


In [26]:
%pip install boto3


     -------------------------------------- 135.5/135.5 kB 1.0 MB/s eta 0:00:00
     ---------------------------------------- 10.5/10.5 MB 7.1 MB/s eta 0:00:00
     ---------------------------------------- 79.6/79.6 kB 4.3 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [28]:
import boto3
from contextlib import closing

def text_to_speech_amazon_polly(text, output_file):
    polly_client = boto3.Session(region_name='us-east-1').client('polly')
    response = polly_client.synthesize_speech(Text=text, OutputFormat='mp3', VoiceId='Joanna')

    with closing(response['AudioStream']) as stream:
        with open(output_file, 'wb') as file:
            file.write(stream.read())

text = "This is a sample text for converting to speech using Amazon Polly."
output_file = "output_amazon_polly.mp3"
text_to_speech_amazon_polly(text, output_file)


NoCredentialsError: Unable to locate credentials

In [29]:
%pip install pydub


Note: you may need to restart the kernel to use updated packages.


In [31]:
from pydub import AudioSegment

def concatenate_audio_files(file1, file2, output_file):
    # Load audio files
    audio1 = AudioSegment.from_mp3(file1)
    audio2 = AudioSegment.from_mp3(file2)

    # Concatenate audio files
    combined_audio = audio1 + audio2

    # Export the result
    combined_audio.export(output_file, format="mp3")

file1 = "output_gtts.mp3"
file2 = "output_pyttsx3.mp3"
output_file = "output_concatenated.mp3"
concatenate_audio_files(file1, file2, output_file)


c:\Users\souvik.dey\AppData\Local\Programs\Python\Python38\lib\site-packages\pydub\utils.py:198: RuntimeWarning: Couldn't find ffprobe or avprobe - defaulting to ffprobe, but may not work
  warn("Couldn't find ffprobe or avprobe - defaulting to ffprobe, but may not work", RuntimeWarning)


FileNotFoundError: [WinError 2] The system cannot find the file specified

In [1]:
%pip install tk


Note: you may need to restart the kernel to use updated packages.


In [2]:
import tkinter as tk

def on_button_click():
    print("Button clicked")

root = tk.Tk()
root.title("Tkinter Sample")

button = tk.Button(root, text="Click me!", command=on_button_click)
button.pack()

root.mainloop()


Button clicked


In [3]:
%pip install PyQt5


Note: you may need to restart the kernel to use updated packages.


In [4]:
import sys
from PyQt5.QtWidgets import QApplication, QWidget, QVBoxLayout, QPushButton

def on_button_click():
    print("Button clicked")

app = QApplication(sys.argv)

window = QWidget()
window.setWindowTitle("PyQt Sample")

layout = QVBoxLayout()

button = QPushButton("Click me!")
button.clicked.connect(on_button_click)

layout.addWidget(button)

window.setLayout(layout)
window.show()

sys.exit(app.exec_())


Button clicked
Button clicked
Button clicked
Button clicked
Button clicked


SystemExit: 0

C:\Users\souvik.dey\AppData\Roaming\Python\Python38\site-packages\IPython\core\interactiveshell.py:3386: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


In [5]:
%pip install kivy


     ---------------------------------------- 4.0/4.0 MB 4.0 MB/s eta 0:00:00
     ---------------------------------------- 4.8/4.8 MB 7.1 MB/s eta 0:00:00
     ---------------------------------------- 3.1/3.1 MB 10.3 MB/s eta 0:00:00
     ------------------------------------- 570.5/570.5 kB 11.9 MB/s eta 0:00:00
     -------------------------------------- 123.6/123.6 kB 3.5 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [6]:
from kivy.app import App
from kivy.uix.button import Button

class MyApp(App):
    def on_button_click(self, instance):
        print("Button clicked")

    def build(self):
        button = Button(text="Click me!")
        button.bind(on_press=self.on_button_click)
        return button

if __name__ == '__main__':
    MyApp().run()


[WARNING] [Config      ] Older configuration version detected (0 instead of 24)
[WARNING] [Config      ] Upgrading configuration in progress.
[INFO   ] [Logger      ] Record log in C:\Users\souvik.dey\.kivy\logs\kivy_23-03-28_0.txt
[ERROR  ] [Core        ] option --ip not recognized


Kivy Usage: ipykernel_launcher.py [KIVY OPTION...] [-- PROGRAM OPTIONS]::

            Options placed after a '-- ' separator, will not be touched by kivy,
            and instead passed to your program.

            Set KIVY_NO_ARGS=1 in your environment or before you import Kivy to
            disable Kivy's argument parser.

        -h, --help
            Prints this help message.
        -d, --debug
            Shows debug log.
        -a, --auto-fullscreen
            Force 'auto' fullscreen mode (no resolution change).
            Uses your display's resolution. This is most likely what you want.
        -c, --config section:key[:value]
            Set a custom [section] key=value in the configuration object.
        -f, --fullscreen
            Force running in fullscreen mode.
        -k, --fake-fullscreen
            Force 'fake' fullscreen mode (no window border/decoration).
            Uses the resolution specified by width and height in your config.
        -w, --windowed
 

AssertionError: 

In [1]:
%pip install pdfplumber

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pdfplumber

def read_pdf(file_path):
    with pdfplumber.open(file_path) as pdf:
        pages = [page.extract_text() for page in pdf.pages]
    return pages

file_path = "srim.pdf"
pages = read_pdf(file_path)
print(type(pages))

<class 'list'>


In [3]:
print(pages)

['', "Forthcoming Titles from Magenta Press\nMountains, Planes, Barren Deserts, and What Lies Beyond -- The Journey Continues... by Sri M,\nAuthor of Apprenticed to a Himalayan Master: A Yogi's Autobiography.\nHow to Levitate and Other Secrets of Magic by James Talbot\nWisdom of the Great Sages: Conversations with Sri Guru, Babaji and Other Masters\nOther Titles by Sri M\nThe Little Guide to Greater Glory and a Happier Life, published by Hima Communication\nWisdom of the Rishis: The Three Upanishads: Ishavasya - Kena - Mandukya, published by Satsang\nCommunications\nNew edition of Jewel in the Lotus: Deeper Aspects of Hinduism, published by Magenta Press", '', '© The author, 2010\nAll rights reserved. No part of this publication may be reproduced, stored in retrieval system or\ntransmitted by any means, electronic, photocopying, recording, or otherwise, without the prior written\npermission of the copyright holders.\neISBN: 978-93-82585-86-2\nFirst printed December 2010\nFourth impress

In [4]:
def get_chapters_list(pages, index_start, index_finish):
    return pages[index_start-1:index_finish-1]

index_start = 5 # Adjust this value to the page where the index starts
index_finish = 7 # Adjust this value to the page where the index ends
index_pages = get_chapters_list(pages, index_start, index_finish)
print(type(index_pages))

<class 'list'>


In [5]:
# print(index_pages)
# print(type(index_pages[0]))

for val in index_pages:
    print(type(val))
    # Split the string by the newline character
    my_list = val.split('\n')

    # Print the list
    print(my_list)

<class 'str'>
['Contents', 'Title Page', 'Copyright & Permissions', 'Acknowledgements', 'Foreword', 'List of Photo Inserts', '1. The Beginning', '2. A Visit from the Himalayan Master', '3. The Gayatri Mantra', '4. Meeting Yogi Gopala Saami', '5. A God-Intoxicated Sufi', '6. Inside the Subramanya Temple', '7. Preparing for the Ascent', '8. The Story of Sri Narayana Guru', '9. Strange and Timely Inputs', '10. The Mastan’s Blessings', '11. Three Monks', '12. Mai Ma', '13. Understanding Holy Madness', '14. Trial Run', '15. Towards the Himalayas', '16. The Cave of Vasishta', '17. The Naga and Learning the Mahamantra', '18. The Cobbler and the Monk', '19. The Seeker at Badrinath', '20. Meeting Babaji']
<class 'str'>
['21. My First Initiation', '22. Lighting the Kundalini Fire', '23. The Old Tibetan Lama', '24. Yoga, Vedanta, and the Nath Pant', '25. Tholingmutt and the Mountain Man', '26. The Valley of Flowers and Hemkund', '27. Meeting the Siddhar', '28. Kedarnath: Opening the Channels', '2

In [7]:
def filter_chapters(sentences):
    filtered_sentences = []
    for sentence in sentences:
        first_word = sentence.split()[0]
        if first_word[0].isdigit():
            filtered_sentences.append(sentence.lower())
    return filtered_sentences

filtered_chapters_list = []
for val in index_pages:
    val = val.split('\n')
    temp_chapter_list = filter_chapters(val)
    filtered_chapters_list.extend(temp_chapter_list)
print(filtered_chapters_list)


['1. the beginning', '2. a visit from the himalayan master', '3. the gayatri mantra', '4. meeting yogi gopala saami', '5. a god-intoxicated sufi', '6. inside the subramanya temple', '7. preparing for the ascent', '8. the story of sri narayana guru', '9. strange and timely inputs', '10. the mastan’s blessings', '11. three monks', '12. mai ma', '13. understanding holy madness', '14. trial run', '15. towards the himalayas', '16. the cave of vasishta', '17. the naga and learning the mahamantra', '18. the cobbler and the monk', '19. the seeker at badrinath', '20. meeting babaji', '21. my first initiation', '22. lighting the kundalini fire', '23. the old tibetan lama', '24. yoga, vedanta, and the nath pant', '25. tholingmutt and the mountain man', '26. the valley of flowers and hemkund', '27. meeting the siddhar', '28. kedarnath: opening the channels', '29. the fireball from the sky', '30. the healing and meeting the grand master', '31. a truly holy man', '32. sri vidya initiation', '33. les

In [8]:
def filter_initial_pages(pages, start_page):
    return pages[start_page-1:]

start_page = 15 # Adjust this value to the page where the main content starts
filtered_pages = filter_initial_pages(pages, start_page)
print(type(filtered_pages))

<class 'list'>


In [9]:
print(len(filtered_pages))

248


In [10]:
all_pages_text = []
for val in filtered_pages:
    page_text = val.split('\n')
    all_pages_text.append(page_text)

print(all_pages_text)
print(len(all_pages_text))

[['1. The Beginning', 'Forty', 'years ago, a young man of nineteen from Kerala, the southern-most state of India, was', 'found in deep meditation in the cave of Vyasa up in the Himalayas near Badrinath, close to the Indo-', 'Tibetan border regions. Even in those days, a teenager going away to the Himalayas to meditate was', 'not a common occurrence. What made this case almost unbelievable was that the young man in', 'question was not even born a Hindu.', 'How this young man became a yogi; and the mysterious and fascinating world of un-imaginable', 'power and grandeur that opened up for him, is the story of my life. I was this young man.', 'If you will allow me, I shall start at the very beginning. Let us, as quickly as possible, walk', 'together to the abode of the eternal snows. There, among the ice-clad Himalayan peaks, dwelt my', 'friend, philosopher and guide – my dearest teacher and master by whose kindness and grace, I learnt', 'to fly high into the greater dimensions of consciou

In [44]:
# # filtered_chapters_list - contains the list of all the chapters 

# def split_into_chapters(filtered_chapters_list, all_pages_text):
#     chapter_wise_text = []
#     triage_page = False 
#     for val in all_pages_text:
#         if val in filtered_chapters_list: 
#             if triage_page == False:
#                 continue
#             else : 


    



In [1]:
cleaned_pages = [page for page in filtered_pages]
print(cleaned_pages)

NameError: name 'filtered_pages' is not defined

In [13]:
print(all_pages_text)

[['1. The Beginning', 'Forty', 'years ago, a young man of nineteen from Kerala, the southern-most state of India, was', 'found in deep meditation in the cave of Vyasa up in the Himalayas near Badrinath, close to the Indo-', 'Tibetan border regions. Even in those days, a teenager going away to the Himalayas to meditate was', 'not a common occurrence. What made this case almost unbelievable was that the young man in', 'question was not even born a Hindu.', 'How this young man became a yogi; and the mysterious and fascinating world of un-imaginable', 'power and grandeur that opened up for him, is the story of my life. I was this young man.', 'If you will allow me, I shall start at the very beginning. Let us, as quickly as possible, walk', 'together to the abode of the eternal snows. There, among the ice-clad Himalayan peaks, dwelt my', 'friend, philosopher and guide – my dearest teacher and master by whose kindness and grace, I learnt', 'to fly high into the greater dimensions of consciou

In [16]:
print(filtered_chapters_list)
for val in filtered_chapters_list:
    print(val)

['1. the beginning', '2. a visit from the himalayan master', '3. the gayatri mantra', '4. meeting yogi gopala saami', '5. a god-intoxicated sufi', '6. inside the subramanya temple', '7. preparing for the ascent', '8. the story of sri narayana guru', '9. strange and timely inputs', '10. the mastan’s blessings', '11. three monks', '12. mai ma', '13. understanding holy madness', '14. trial run', '15. towards the himalayas', '16. the cave of vasishta', '17. the naga and learning the mahamantra', '18. the cobbler and the monk', '19. the seeker at badrinath', '20. meeting babaji', '21. my first initiation', '22. lighting the kundalini fire', '23. the old tibetan lama', '24. yoga, vedanta, and the nath pant', '25. tholingmutt and the mountain man', '26. the valley of flowers and hemkund', '27. meeting the siddhar', '28. kedarnath: opening the channels', '29. the fireball from the sky', '30. the healing and meeting the grand master', '31. a truly holy man', '32. sri vidya initiation', '33. les

In [17]:
import os
import re


def split_into_chapters_by_name(cleaned_sentences, chapter_names):
    chapters = []
    chapter_content = []
    
    for sentence in cleaned_sentences:
        sentence = [s.lower() for s in sentence]
        for name in chapter_names:
            if name in sentence:
                if chapter_content:
                    chapters.append(chapter_content)
                chapter_content = []
                break
        chapter_content.extend(sentence)
    
    if chapter_content:
        chapters.append(chapter_content)
    
    return chapters


def save_chapters_to_files(chapters, chapter_names, output_folder):
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)

    for idx, (chapter, name) in enumerate(zip(chapters, chapter_names)):
        file_name = f"{idx+1}_{name}.txt"
        file_path = os.path.join(output_folder, file_name)
        
        with open(file_path, "w", encoding="utf-8") as output_file:
            print(chapter)
            delimiter = " "
            temp_val = delimiter.join(chapter)
            output_file.write(temp_val)

# # A list of chapter names, you should replace these with the actual chapter names from your ebook
# chapter_names = ["Chapter One: The Beginning", "Chapter Two: The Journey", "Chapter Three: The End"]

chapters = split_into_chapters_by_name(all_pages_text, filtered_chapters_list)
# for val in chapters:
#     print(val)
output_folder = "output_chapters"
save_chapters_to_files(chapters, filtered_chapters_list, output_folder)


['1. the beginning', 'forty', 'years ago, a young man of nineteen from kerala, the southern-most state of india, was', 'found in deep meditation in the cave of vyasa up in the himalayas near badrinath, close to the indo-', 'tibetan border regions. even in those days, a teenager going away to the himalayas to meditate was', 'not a common occurrence. what made this case almost unbelievable was that the young man in', 'question was not even born a hindu.', 'how this young man became a yogi; and the mysterious and fascinating world of un-imaginable', 'power and grandeur that opened up for him, is the story of my life. i was this young man.', 'if you will allow me, i shall start at the very beginning. let us, as quickly as possible, walk', 'together to the abode of the eternal snows. there, among the ice-clad himalayan peaks, dwelt my', 'friend, philosopher and guide – my dearest teacher and master by whose kindness and grace, i learnt', 'to fly high into the greater dimensions of conscious

In [ ]:
import os
import re

def split_into_chapters(filtered_pages):
    # # Adjust the pattern to match the chapter headings in your ebook
    # pattern = r'\bChapter\s+\d+\b'
    # chapters = re.split(pattern, filtered_pages)
    # return chapters[1:]  # Exclude the text before the first chapter
    all_pages_text = []
    for val in filtered_pages:
        page_text = val.split('\n')
        all_pages_text.append(page_text)

    print(all_pages_text)
    return all_pages_text

def save_chapters_to_files(chapters, output_folder):
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)

    for idx, chapter in enumerate(chapters, start=1):
        file_name = f"Chapter_{idx}.txt"
        file_path = os.path.join(output_folder, file_name)
        
        with open(file_path, "w", encoding="utf-8") as output_file:
            output_file.write(chapter)

chapters = split_into_chapters(filtered_pages)
output_folder = "output_chapters"
save_chapters_to_files(chapters, output_folder)


In [18]:
import pyttsx3

def text_to_speech_pyttsx3(text, output_file):
    engine = pyttsx3.init()
    engine.save_to_file(text, output_file)
    engine.runAndWait()

text = "This is a sample text for converting to speech using pyttsx3."
output_file = "output_pyttsx3.mp3"
text_to_speech_pyttsx3(text, output_file)

In [19]:
%pip install pydub ffmpeg

  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Running setup.py install for ffmpeg: started
  Running setup.py install for ffmpeg: finished with status 'done'
Note: you may need to restart the kernel to use updated packages.


  DEPRECATION: ffmpeg is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559


In [20]:
import os
import pyttsx3
from pydub import AudioSegment

def create_output_directory(directory):
    if not os.path.exists(directory):
        os.makedirs(directory)

def get_text_files(input_folder):
    return [file for file in os.listdir(input_folder) if file.endswith(".txt")]

def read_text_file(file_path):
    with open(file_path, "r", encoding="utf-8") as text_file:
        return text_file.read()

def text_to_speech(text, temp_file="temp.wav"):
    engine = pyttsx3.init()
    engine.setProperty("rate", 150)  # Adjust the speech rate if necessary

    engine.save_to_file(text, temp_file)
    engine.runAndWait()

def convert_wav_to_mp3(wav_file, mp3_file):
    wav_audio = AudioSegment.from_wav(wav_file)
    wav_audio.export(mp3_file, format="mp3")

def process_text_files(input_folder, output_folder):
    create_output_directory(output_folder)
    text_files = get_text_files(input_folder)

    for file in text_files:
        input_file_path = os.path.join(input_folder, file)
        output_file_path = os.path.join(output_folder, file.replace(".txt", ".mp3"))

        text = read_text_file(input_file_path)
        text_to_speech(text)
        convert_wav_to_mp3("temp.wav", output_file_path)

        os.remove("temp.wav")

input_folder = "output_chapters"
output_folder = "output_speech"
process_text_files(input_folder, output_folder)


c:\Users\souvik.dey\AppData\Local\Programs\Python\Python38\lib\site-packages\pydub\utils.py:170: RuntimeWarning: Couldn't find ffmpeg or avconv - defaulting to ffmpeg, but may not work
  warn("Couldn't find ffmpeg or avconv - defaulting to ffmpeg, but may not work", RuntimeWarning)


FileNotFoundError: [WinError 2] The system cannot find the file specified

In [25]:
import os
from gtts import gTTS
import time

def create_output_directory(directory):
    if not os.path.exists(directory):
        os.makedirs(directory)

def get_text_files(input_folder):
    return [file for file in os.listdir(input_folder) if file.endswith(".txt")]

def read_text_file(file_path):
    with open(file_path, "r", encoding="utf-8") as text_file:
        return text_file.read()

def text_to_speech_gtts(text, output_file_path, lang="en"):
    tts = gTTS(text, lang=lang)
    tts.save(output_file_path)

def process_text_files(input_folder, output_folder):
    create_output_directory(output_folder)
    text_files = get_text_files(input_folder)

    for file in text_files:
        input_file_path = os.path.join(input_folder, file)
        output_file_path = os.path.join(output_folder, file.replace(".txt", ".mp3"))

        text = read_text_file(input_file_path)
        text_to_speech_gtts(text, output_file_path)
        time.sleep(300)

input_folder = "output_chapters"
output_folder = "output_speech"
process_text_files(input_folder, output_folder)


gTTSError: 429 (Too Many Requests) from TTS API. Probable cause: Unknown

Challenge faced is  this :
Throttling issue with gtts
 
Output exceeds the size limit. Open the full output data in a text editor---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
File c:\Users\souvik.dey\AppData\Local\Programs\Python\Python38\lib\site-packages\gtts\tts.py:269, in gTTS.stream(self)
    267     log.debug("status-%i: %s", idx, r.status_code)
--> 269     r.raise_for_status()
    270 except requests.exceptions.HTTPError as e:  # pragma: no cover
    271     # Request successful, bad response

File c:\Users\souvik.dey\AppData\Local\Programs\Python\Python38\lib\site-packages\requests\models.py:1021, in Response.raise_for_status(self)
   1020 if http_error_msg:
-> 1021     raise HTTPError(http_error_msg, response=self)

HTTPError: 429 Client Error: Too Many Requests for url: https://www.google.com/sorry/index?continue=https://translate.google.com/_/TranslateWebserverUi/data/batchexecute&q=EgRnmupKGOKGp6EGIjBtAoFPTeRRN1nS1ej7suG2pY-JP_EpYr2M1S-H_X1ieCL0a8XTE2Tor4unzzUZrWUyAXI

During handling of the above exception, another exception occurred:

gTTSError                                 Traceback (most recent call last)
c:\Users\souvik.dey\Desktop\AutomationCode\GitHub\Self_RnD\frugal_audiobook.ipynb Cell 41 in 3
     32 input_folder = "output_chapters"
     33 output_folder = "output_speech"
---> 34 process_text_files(input_folder, output_folder)

c:\Users\souvik.dey\Desktop\AutomationCode\GitHub\Self_RnD\frugal_audiobook.ipynb Cell 41 in 2
     26 output_file_path = os.path.join(output_folder, file.replace(".txt", ".mp3"))
     28 text = read_text_file(input_file_path)
...
    274 except requests.exceptions.RequestException as e:  # pragma: no cover
    275     # Request failed
    276     log.debug(str(e))

gTTSError: 429 (Too Many Requests) from TTS API. Probable cause: Unknown